DM-37584: add proposal from the Great Calibration Refactor discussions

TallJimbo opened the pull request.

TallJimbo committed:
Add new section and diagram for Great Calibration Refactor.

TallJimbo committed:
Updates during paircoding work.

TallJimbo committed:
Finish first draft of Great Calibration Refactor proposal.

leeskelvin commented:
Thanks Jim and Eli for a great talk today! From reading through your proposals, here are my initial thoughts on the refactor:
* I'm concerned that merging of many smaller PipelineTasks into a small number of "SuperPipelineTasks" may limit the discoverability of exactly what data reduction is happening within a task. The concept of SuperPipelineTasks  to my eye seems somewhat analogous to the use of subsets within a pipeline definition file, however, having data reduction grouped into a smaller number of SuperPipelineTasks limits the flexibility of subsets (and to some extent, replaces the need for subsets). If anything, I'd actually prefer to break up some of the larger PipelineTasks up into smaller tasks, and then rely on subsets to link them together for the end-user to interact with. However, if the win from I/O optimization is significant, then I can see the logic to going this way instead of leaning so heavily on subsets.
* I agree with/fully support pretty much all of the proposed naming conventions, except for:
  * I would prefer all table types to have an appropriate suffix, in contrast to your proposal that "the final thing does not need a suffix". I.e., I would prefer to see `initial_source_visit` instead of the proposed `initial_source`. This makes it clearer immediately as to what level of data processing this initial source table belongs to (and also saves us from any future unexpected 'initial_source' datasets which we may want to construct). This logic may also be advocating for `unforced_source` in place of just `source`, but that may not be necessary.
  * The use of `initial_` and `final_` for certain dataset type names inherently locks-in these dataset types as being the 'initial' and 'final' versions of these data. If this is really our last big change to the data reduction pipeline, then this may be fine. However, for example, if at some future point we add a new task that sits after `bootstrapImage` that produces an updated `initial_pvi` output, I don't see how we could appropriately name this output unless we go down the route of switching the existing `initial_pvi` to an `initial_pre_pvi`, or similar. I personally prefer how `isr` does this right now, using a `post...` syntax for output data products, but I accept that that may not be desirable for long task name outputs. However, `initial_` and `final_` may not be important considerations in the first place, as I note that you don't intend any of these data products to be persisted on-disk (so this may all be an internal feature and therefore just an implementation detail). You may also argue that any new task should instead be merged into `bootstrapImage`, removing the problem outlined here, which is also a valid point.
* The term `bootstrap` to me does not seem to convey what these proposed super tasks are doing (I instead think of bootstrapped errors). I like the idea of reusing `characterizeImage` here, but I accept this may not be desirable if we wish to remove any possibility for confusion post-refactor. I think John mentioned that `calibrateImage` is also available, which may work well too.

timj commented:
I know nothing about this discussion but:

> if the win from I/O optimization is significant,

The idea from the beginning was that for some cases where we go from one pipeline task to another and it's clustered into a single process and we don't need the intermediate, we would use the in-memory datastore and there wouldn't be any I/O. This design makes it very easy to configure datastore to allow those intermediates to be persisted for debugging if you want to do that.

natelust commented:
I was going to chime in and say the same thing as @timj . I think since we were talking about middleware changes being needed anyway, we should put that effort into the clustering algorithm / in-memory datastore instead. I understand the point of a small change to characterize/calibrate possibly being appropriate, but as a generality the more tightly coupled larger tasks are, the harder it is to possibly change the behavior in the future, by inserting things in between stages. The pipeline system was not just a system to manage batch production issues, but also a tool to be leveraged by developers.

All this said I just wanted to chime in on that discussion above. I still need to finish reading through the proposal, and will have more detailed thoughts later.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+- Task labels are camelCase and start with a lowercase verb: "associateIsolatedStars" instead of "isolatedStarAssociation".
+- Dataset type names are snake_case nouns preceded (if necessary) by adjectives: e.g. "initial_visit_summary"
+- Use "source" instead of "src" or "sources".
+- Avoid "catalog" or "cat" in dataset type names; use "source", "stars", "object", or "matches" instead when appropriate.
```
I am split on this, I agree that if you are in the know, it does not add any additional information, but if you are not and just read a list of data set types, seeing `stars` does not really tell me anything. Is it a stack of postage stamps of isolated stars? Is it a data set type used in creation of synthetic stars? At least if there is a `table` or `catalog` somewhere in there it makes it a bit clearer what it is.

TallJimbo replied:
I somewhat agree w.r.t. `stars`, but I do think `object`, `source`, and `matches` pretty clearly imply table/catalog.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+- Dataset type names are snake_case nouns preceded (if necessary) by adjectives: e.g. "initial_visit_summary"
+- Use "source" instead of "src" or "sources".
+- Avoid "catalog" or "cat" in dataset type names; use "source", "stars", "object", or "matches" instead when appropriate.
+- Use some variant of "pvi" for any direct (non-difference) ``{visit, detector}`` image dataset with an image, mask, and variance plane (i.e. ``lsst.afw.image.Exposure`` or ``MaskedImage``).
```
Im not sure reusing the same `pvi` name for two different python dataset types is the best idea, it may be confusing when someone is looking to load something into a notebook.

TallJimbo replied:
I'm not sure what you mean.  These aren't Python things, and it's no different from appending "exp" to "calexp" or "icExp" now (just better, I maintain).

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+ISR and coaddition will also see major changes (spurred by Calibpalooza and cell-based and chi-squared coadds, respectively) on similar timescales (we hope), but we're considering those out-of-scope and mostly orthogonal.
+While we can and should implement this piecemeal when we can, we want a complete vision of what it will look like in the end, and in some cases it may be easier to replace many tasks at once.
+
+We lean towards merging PipelineTasks with the same dimensions that run back-to-back rather than keeping them distinct.
```
I commented on this on the main thread, but I really don't think this is the right approach. I think it would be better to spend the time on the operation side to cluster / in-memory data store. I don't really think it is appropriate at this level for task authors to design their algorithms around optimizing i/o

TallJimbo replied:
I'll add a "when all other considerations are equal" qualifier here, but middleware is way behind on delivering on promises to pipelines already, with some major middleware deliverables still on the horizon.  Pipelines development effort and middleware development effort are not fungible, and even if they were, merging a pair of pretty-similar PipelineTasks (or, at worst, maybe three such pairs - that's really the most we might be debating here) is not a lot of work, and the same cannot be said for in-memory datastore and clustering work.

timj replied:
I tend to agree. The clustering+in-memory data store is pretty close now (especially now that I added support for copy-on-get) but it's all the support code that is lacking such as getting BPS to configure the in-memory datastore per-task when it sees there are intermediates in the clustering, and working out which datasets have to be copied on get and which don't have to be.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+- Avoid "catalog" or "cat" in dataset type names; use "source", "stars", "object", or "matches" instead when appropriate.
+- Use some variant of "pvi" for any direct (non-difference) ``{visit, detector}`` image dataset with an image, mask, and variance plane (i.e. ``lsst.afw.image.Exposure`` or ``MaskedImage``).
+- Tables that are initially per-detector that will get concatenated into per-visit tables should get a "_detector" suffix so the final thing does not need a suffix (and when we revamp later steps of the pipeline, the same for "_patch" so there's no "_tract").
+- Task labels and dataset type names are for "slots", not specific tasks or connections - usually those are 1-1, but when they are not, the label and dataset type names should remain fixed when a different task is swapped in (our "solveAstrometry" task label slot could be satisfied by either jointcal or GBDES).
```
I have long wanted this, and we tried to do it, but it is currently impossible (It will require refactoring middleware)

TallJimbo replied:
In the sense that trying this without middleware changes makes pipelines to hard to import-and-modify to swap out tasks?  Or is there something that really does make it technically impossible?

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+- Use some variant of "pvi" for any direct (non-difference) ``{visit, detector}`` image dataset with an image, mask, and variance plane (i.e. ``lsst.afw.image.Exposure`` or ``MaskedImage``).
+- Tables that are initially per-detector that will get concatenated into per-visit tables should get a "_detector" suffix so the final thing does not need a suffix (and when we revamp later steps of the pipeline, the same for "_patch" so there's no "_tract").
+- Task labels and dataset type names are for "slots", not specific tasks or connections - usually those are 1-1, but when they are not, the label and dataset type names should remain fixed when a different task is swapped in (our "solveAstrometry" task label slot could be satisfied by either jointcal or GBDES).
+  Whether the same is true of the outputs is an open question we'd like to discuss.
```
This point in context with the last is confusing to me, surely the inputs to a task don't really matter (i.e. maybe one takes 3 more prerequisites) but the outputs should remain constant if you expect to substitute it in.

TallJimbo replied:
Yes, I'm not sure what I was thinking here.  I'll delete this sentence.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+- Task labels and dataset type names are for "slots", not specific tasks or connections - usually those are 1-1, but when they are not, the label and dataset type names should remain fixed when a different task is swapped in (our "solveAstrometry" task label slot could be satisfied by either jointcal or GBDES).
+  Whether the same is true of the outputs is an open question we'd like to discuss.
+- "initial" catalogs are not (ever) SDM-standardized, while final catalogs are always SDM-standardized before they are written out.
+  This may present a challenge for analysis tools that want to be able to run on either, but I'm hoping we can minimize analysis of the initial things, as we don't care how good they are in the end.
```
I think this point is just not true, and think it needs much more justification than a simple assertion.

TallJimbo replied:
Agreed; my thinking on this has evolved while watching the PDR2 processing play out.  I'll switch this to a TODO entry, as I want to get this branch merged, but we do need to adjust the proposal to address this.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+  Whether the same is true of the outputs is an open question we'd like to discuss.
+- "initial" catalogs are not (ever) SDM-standardized, while final catalogs are always SDM-standardized before they are written out.
+  This may present a challenge for analysis tools that want to be able to run on either, but I'm hoping we can minimize analysis of the initial things, as we don't care how good they are in the end.
+- Do include a "final" prefix on dataset types that represent the best version of things that are nevertheless temporaries that will not be retained.
```
I agree its good to not include a "final" on the end of something, but assuming that something without a suffix is the final product is also not right. At any point another task may come along and be added to the chain, but we cannot keep updating existing dataset type names. It will really break all downstream workflows

TallJimbo replied:
If we need to change the task that produces a final thing, to a new task that runs after the original one, the right thing to do is to change the dataset type configured for the first one, so the new task can use the final name.

If the dataset types do not have the same storage class or dimensions, that's a much bigger problem because they aren't substitutable.  I agree it could still happen, but I think it'd come with some extra considerations that make these sort of general conventions less relevant anyway.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+  This may present a challenge for analysis tools that want to be able to run on either, but I'm hoping we can minimize analysis of the initial things, as we don't care how good they are in the end.
+- Do include a "final" prefix on dataset types that represent the best version of things that are nevertheless temporaries that will not be retained.
+  Do not add any prefix to dataset types that will be retained for public access.
+- Convert source and object catalogs to Parquet before ever persisting them; only use SourceCatalog/FITS to hold Footprints.
```
Why not store fooprints in a completely orthogonal way?

TallJimbo replied:
I'd be happy to, if we had another way to do it.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+- Do include a "final" prefix on dataset types that represent the best version of things that are nevertheless temporaries that will not be retained.
+  Do not add any prefix to dataset types that will be retained for public access.
+- Convert source and object catalogs to Parquet before ever persisting them; only use SourceCatalog/FITS to hold Footprints.
+  We will retain the ``slot_*`` column names and not the underlying names they point to - downstream code should only be referring to the slot columns anyway, and in many cases we won't run multiple algorithms that could satisfy the slot.
```
Overall, I think we should re-visit the idea of slot altogether. I understand why it was introduced, but I think in practice measurements are not isometric to each other, and most downstream consumers actually care about the algorithm, such that they will always directly access that named column, or slots can never really be changed for how it will disrupt downstream consumers.

TallJimbo replied:
I don't necessarily disagree but I'm considering that out of scope here.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+   Footprints will be written to a separate ``SourceCatalog`` dataset, ``initial_footprints``.
+   Whether this task does multiple detection rounds (to iterate on CR detection, the source detection filter, or the detection threshold) is TBD; we'd like to minimize that.
+   We would like to avoid running a deblender here (if we do, it will have to happen after we have the PSF model), but this depends on how this performs on crowded fields.
+   We are not thrilled with the name of this task and would love ideas for improvements (which should be coordinated with the names for ``bootstrapVisit``, ``finalizeImage``, ``compressImage``, and ``rebuildImage``).
```
I think this misses the legitimate use case of calexp now, that of a "good enough" image for things, even if it is not the final calibration (especially for pipelines that do not run FGCM) but I may not be following how that use case plays out in the proposed graph

TallJimbo replied:
In pipelines that do not produce a better version of the image produced, the image produced here should have a name that does not imply that it will be superseded.  But I want to focus on the actual DRP pipeline here.

natelust commented on index.rst:
```diff
@@ -249,6 +344,11 @@ What are the open questions that drive the differences between different viable
 
+- Do we need some kind of coaddition or warp-comparison to finalize our per-visit background models?  If so, will those also impact PSF modeling and/or aperture corrections?
+
+- Is a second round of astrometric fitting necessary to ensure PSF chromaticity is consistent with our (achromatic) WCSs?  Is it sufficient?
```
Are you trying to soak up dcr with the PSF?

TallJimbo replied:
Yes.  Putting DCR in the WCS is both really problematic (you can't warp images to the same pixel grid for coaddition if the WCS depends on sub-band SED) and at best an approximation (it's not a shift, it's an asymmetric smearing, unless the SED is a delta function), while it can be fully corrected by having a chromatic PSF.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+modelVisitBackground
+   A replacement for SkyCorrectionTask, probably using PCA.
+   SkyCorrectionTask will serve as a placeholder until we have something better.
+   Also takes care of subtracting the wings of bright stars.
```
How do you expect this to handle real astronomical background?

TallJimbo replied:
We can't do background matching here, if that's what you mean; it'd have to go later, and be a part of coaddition.

I'm mostly assuming for now that we won't get background matching implemented for DR1, because nobody's working on it right now, and it's a massive complication to the whole structure here that I don't think we'll be able to nail down until/unless we have something mostly working anyway.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+finalizeCharacterization
+   Nothing new here except new names for the output datasets.
+
+   We are using "characterizations" here to mean "PSFs and aperture corrections", and are not thrilled about that, but we need something that means "PSFs and aperture corrections" that could also absorb aperture corrections being done rather differently than they are today.
```
finalizeStelarProfileCorrections?

TallJimbo replied:
Not a bad idea, but at this point I'm just trying to get this merged and don't have energy to shepherd a naming decision.  Will punt.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+
+finalizeImage
+   This new task takes all of our hard-won final characterizations and calibrations of the image and produces a final ``{visit, detector}`` ``pvi`` image and full-depth, all-measurements ``final_source_detector`` catalog.
+   The latter will be SDM-standardized before it is ever written to disk.
```
I think it is always dangerous to embed final in any term, because you risk the dissertation problem of ``finalizeImageV2Updated``

TallJimbo replied:
But these aren't files or datasets; they're dataset types, and the idea here is that those represent interfaces, not particular task outputs.  So you don't add "V2Updated", you reconfigure the original task that produced `finalizedImage` to write `penultimateImage` (or whatever) instead and let the new final task start writing `finalizedImage`.

natelust commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+   This task just concatenates the ``final_source_detector`` catalogs into a single per-visit ``source`` catalog.
+
+compressImage
+   This task reads the ``pvi`` dataset, lossy-compresses the image and variance planes, lossless-compresses everything else (at least the mask; I don't know if compressing more than that is possible).
```
I am really not sure why you want to do lossy-compression, there are some really impressive lossless compression algorithms. And on that note, what is the average percentage difference in storage space between lossy compressed images vs lossless?

timj replied:
Robert Gruendl did an analysis at some point: https://dmtn-068.lsst.io

Without lossy compression the assertion was that we would not be able to archive the PVIs at all and would have to regenerate them on demand. It was a big struggle to include the PVIs in the costing model at all.

TallJimbo replied:
Echoing Tim here: I don't think there are lossless compression of _floating point_ images that get anywhere near the compression ratios we need to make saving PVIs viable.  Lots of prior art here, but I don't know the details.

TallJimbo committed:
Fix dimensions for bootstrapVisit.

TallJimbo reviewed (commented):
Thanks, all, for your patience.  I haven't made the changes promised in the review responses yet, but my plan is to do that shortly and then get this merged, to allow others to start to make their own changes on top of it.

> I'm concerned that merging of many smaller PipelineTasks into a small number of "SuperPipelineTasks" may limit the discoverability of exactly what data reduction is happening within a task.

My inclusion of "a preference for merging" as a goal certainly kicked a hornet nest (this was a big piece @natelust 's and @timj 's reviews as well), and I think it was read as a much more extreme proposal than I intended; for example:

 - Do we merge `characterizeImage` and `calibrate`?  Yes, they have the same dimensions and an unclear boundary already.
 - Do we merge `skyCorr` and `finalizeCharacterizeation`?  No, they have clearly separate roles and mostly different inputs, even if they have the same dimensions.

So _mostly_ my response to these concerns is that we should resolve cases where this question comes up case-by-case, because we're not as far apart as we might seem.

But I do also think (speaking with my middleware hat on) that middleware has promised an awful lot that we have yet to deliver, and I am not at all confident that we'll will be able to mitigate all of the challenges that come from big, complex pipelines the way we once envisioned, on the timescales that matter.  Compromises will have to be made somewhere.

> I would prefer all table types to have an appropriate suffix, in contrast to your proposal that "the final thing does not need a suffix".

My main concern is for public dataset types visible to science users, who should not have to say "pvi_visit_detector" or "source_visit", especially since they will never see the intermediate dataset types.  So the question is how much our intermediate dataset types mimic those or intentionally differ from them, and my feelings are not nearly as strong about that.

> The use of initial_ and final_ for certain dataset type names inherently locks-in these dataset types as being the 'initial' and 'final' versions of these data.

I think we have different conceptions of what happens when we add a new task that takes on the role of producing the final thing: I think we let that new task _steal_ the existing dataset type name, and reconfigure the previously-final task to use a new dataset type that makes its intermediate status clear.  This is indeed quite different from how we do things now, with task names often baked into dataset type names instead, but I think it will improve our ability to define similar-but-not-identical pipelines.  It will also obfuscate precisely which task produced a thing, but at some level I consider this a feature, not a bug: we want the dataset types to be the stable interfaces, not the tasks.

> The term bootstrap to me does not seem to convey what these proposed super tasks are doing (I instead think of bootstrapped errors).

You're not alone, and people who care more than I do have settled on `CalibrateImage`, I think.

TallJimbo committed:
Qualify preference for fewer PipelineTasks.

TallJimbo committed:
Rewrite expections regarding SDM standardization of initial catalogs.

Not running analysis tooling on these was not realistic, but that
doesn't mean we need to fully standardize them, either.

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+- Dataset type names are snake_case nouns preceded (if necessary) by adjectives: e.g. "initial_visit_summary"
+- Use "source" instead of "src" or "sources".
+- Avoid "catalog" or "cat" in dataset type names; use "source", "stars", "object", or "matches" instead when appropriate.
+- Use some variant of "pvi" for any direct (non-difference) ``{visit, detector}`` image dataset with an image, mask, and variance plane (i.e. ``lsst.afw.image.Exposure`` or ``MaskedImage``).
```
Is the plan to replace these `*_pvi` datasets with the replacement for `Exposure` once it becomes available, or to make a new dataset to hold the new type? I'm not clear on whether the new object will be functionally a drop-in replacement.

TallJimbo replied:
At this point I don't have enough confidence that the RFC-950 `Exposure`-replacement will happen to take any action on that basis.  Maybe we'll actually manage to migrate the dataset type definitions, or maybe we'll just hold our noses and add a prefix or suffix for the new ones.

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+  It may be necessary to (partially?) SDM-standardize initial catalogs in order to feed them into the same analysis tooling we expect to run on final catalogs (which will certainly be necessary for at least per-step pipeline validation during full-scale data release processing).
+  But we prefer to consider these initial standardization processes part of the "analysis addition" to the pipeline, and keep it out of the main production pipeline flow, and to minimize this standardization in favor of configuring the input column expected by analysis tasks.
+- Do include a "final" prefix on dataset types that represent the best version of things that are nevertheless temporaries that will not be retained.
+  Do not add any prefix to dataset types that will be retained for public access.
```
I think this should be reworded in the other order; the important thing here is that public-facing outputs are just named "X", while the things that go into those outputs are named "final_Y". I think flipping the order will make that more clear.

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+- Do include a "final" prefix on dataset types that represent the best version of things that are nevertheless temporaries that will not be retained.
+  Do not add any prefix to dataset types that will be retained for public access.
+- Convert source and object catalogs to Parquet before ever persisting them; only use SourceCatalog/FITS to hold Footprints.
+  We will retain the ``slot_*`` column names and not the underlying names they point to - downstream code should only be referring to the slot columns anyway, and in many cases we won't run multiple algorithms that could satisfy the slot.
```
Retaining slot names in the parquet tables seems like a dangerous choice to me: there's no way to know what actually went into that slot (unlike the aliases in afw) and the name itself doesn't tell you. And that can really matter: see e.g. slot_apFlux for HSC vs. LATISS!

TallJimbo replied:
I agree there's a provenance problem (you _should_ be able to figure it out from the configs, but that's at best a pain).  But if we drop the slot names everything downstream gets more fragile, because they have to effectively bake in upstream configuration to figure out which field name they want.  This still feels like the least bad option, though I concede it's not actually a good one.

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+Task and dataset type notes
+"""""""""""""""""""""""""""
+
+bootstrapImage
```
Recommend a global change:

```suggestion
calibrateImage
```

And change the dotfile and SVG.

TallJimbo replied:
Oops, I actually had this change sitting, unpushed, in my local git repo already.

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+
+bootstrapImage
+   Initial background subtraction and detection of bright stars (galaxies are considered a nuisance here).
+   Initial versions of everything - astrometry, photometry, probably PSFs.
```
```suggestion
   Initial versions of everything - background, astrometry, photometry, PSFs.
```

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+   Initial versions of everything - astrometry, photometry, probably PSFs.
+   Probably aperture corrections of some kind, but targeted specifically at making compensated apertures work for FGCM.
+   These are all attached to its image output, ``initial_pvi``, which is a lot like today's ``calexp``.
+   This will be a fluence image with nJy pixel units, but how much of that is done by this task vs. ISR is TBD.
```
```suggestion
   This will be a fluence image with nJy pixel units, but the details of the fluence (raw)->fluence (postISR) conversion will be mostly left to how we create our flat fields and how they are applied in ISR.
```

Or something like that. At least, I think this is what you were getting at here?

TallJimbo replied:
I'll just strike the second clause entirely.  It doesn't add any information, just uncertainty.

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+   Probably aperture corrections of some kind, but targeted specifically at making compensated apertures work for FGCM.
+   These are all attached to its image output, ``initial_pvi``, which is a lot like today's ``calexp``.
+   This will be a fluence image with nJy pixel units, but how much of that is done by this task vs. ISR is TBD.
+   Its output catalog, ``initial_source_detector``, will be converted to Parquet before it is written, but not SDM-standardized.
```
```suggestion
   Its output catalog, ``initial_stars_detector``, will be converted to Parquet before it is written, but not SDM-standardized.
```

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+   These are all attached to its image output, ``initial_pvi``, which is a lot like today's ``calexp``.
+   This will be a fluence image with nJy pixel units, but how much of that is done by this task vs. ISR is TBD.
+   Its output catalog, ``initial_source_detector``, will be converted to Parquet before it is written, but not SDM-standardized.
+   Footprints will be written to a separate ``SourceCatalog`` dataset, ``initial_footprints``.
```
```suggestion
   Footprints will be written to a separate ``SourceCatalog`` dataset, ``initial_stars_footprints_detector``.
```

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+   Footprints will be written to a separate ``SourceCatalog`` dataset, ``initial_footprints``.
+   Whether this task does multiple detection rounds (to iterate on CR detection, the source detection filter, or the detection threshold) is TBD; we'd like to minimize that.
+   We would like to avoid running a deblender here (if we do, it will have to happen after we have the PSF model), but this depends on how this performs on crowded fields.
+   We are not thrilled with the name of this task and would love ideas for improvements (which should be coordinated with the names for ``bootstrapVisit``, ``finalizeImage``, ``compressImage``, and ``rebuildImage``).
```
This sentence can be removed: I think we're happy with `calibrateImage`.

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+   We are not thrilled with the name of this task and would love ideas for improvements (which should be coordinated with the names for ``bootstrapVisit``, ``finalizeImage``, ``compressImage``, and ``rebuildImage``).
+
+bootstrapVisit
+   Consolidate the per-detector outputs of ``bootstrapImage`` and recover from failures on some detectors by using those that succeeded (especially for WCSs).
```
Maybe `mergeDetectorVisits`? It's not `calibrateVisit`, since that happens in fgcm/gbdes.

Or just `consolidateVisit`?

TallJimbo replied:
Ah, see, this is where `bootstrap` was helpful...

But that ship has indeed sailed. I'll go with `consolidateVisit` for now.

parejkoj commented on index.rst:
```diff
@@ -240,6 +240,101 @@ fitSrcMatchAstrometry
+   SkyCorrectionTask will serve as a placeholder until we have something better.
+   Also takes care of subtracting the wings of bright stars.
+
+finalizeCharacterization
+   Nothing new here except new names for the output datasets.
```
Since we no longer have `characterizeImage`, should we rename this? Looks like the existing task produces psf models and apcorr; if we switch to the compensated gaussian filters, we won't need apcorr any more, so could this just be `finalizePsf`?

TallJimbo replied:
We will still need aperture corrections here, even if we can avoid needing them earlier.  I'm happy to rename new tasks (or even substantially changed ones) on this proposal, but I think this one does end up staying pretty much as it is, and I'm inclined to leave its name, too.

parejkoj commented on index.rst:
```diff
@@ -249,6 +344,11 @@ What are the open questions that drive the differences between different viable
+
+- Is a second round of astrometric fitting necessary to ensure PSF chromaticity is consistent with our (achromatic) WCSs?  Is it sufficient?
+
+- Are there small changes we could make to better share code with AP?
```
I believe `CalibrateImageTask` is the primary sharing point: so long as the inputs to that are fluence images with decent camera geometry models that we can fit an affine WCS on top of, the `initial_pvi` should be decent.

After that, it would be great if we could easily make a "get good coadds and enough information to vet them with DRP-like tools" pipeline (our current template generation pipeline is not as sophisticated as DRP's coadd building). Obviously, we want to share as much of the image subtraction code as possible, but I think that's going to happen anyway.

TallJimbo replied:
I agree and I'll go ahead and delete the bullet point.

As for coadds, aside from Y1 _incremental_ templates I think it's probably a subset of a real DRP pipeline you want, not a new DRP-like pipeline.  And while it's not something we have wired together yet, I think eventually you should see that pipeline as producing a deep reference catalog for AP to consume as well as templates (including - but not only - to get object associations for alerts).  That means going a bit further than coaddition, but there's probably still enough in the DRP pipelines after that that a subset would still be worthwhile.

parejkoj reviewed (approved):
A variety of comments; do with them as you will.

TallJimbo committed:
More review responses.
